src/Oscoin/P2P/Handshake/Simple.hs

```diff
@@ -108,17 +110,29 @@ mapRecvError = \case
     Transport.RecvConnReset -> Gone
 
 guardPeerId
-    :: forall i o n c. (Eq n, Serialise (Crypto.PublicKey c))
-    => Proxy c
+    :: forall i o n c proxy. (Eq n, Serialise (Crypto.PublicKey c))
+    => proxy c
     -> Maybe n
     -> HandshakeResult i o n
     -> HandshakeT SimpleError IO (HandshakeResult i o n)
-guardPeerId Proxy (Just peerId) res | peerId /= hrPeerId res = do
+guardPeerId _ (Just peerId) res | peerId /= hrPeerId res = do
     handshakeSend (Bai IdMismatch :: HandshakeMessage c)
     throwError IdMismatch
 
 guardPeerId _ _ res = pure res
 
+guardPeerIdNot
+    :: forall i o n c proxy. (Eq n, (Serialise (Crypto.PublicKey c)))
+    => proxy c
+    -> n
+    -> HandshakeResult i o n
+    -> HandshakeT SimpleError IO (HandshakeResult i o n)
+guardPeerIdNot _ peerId res | peerId == hrPeerId res = do
```

Comment:
methinks -- what about (pseudo code, won't compile):

``` haskell
guardPeer :: Proxy -> n -> (HashshakeResult -> n -> Bool) -> SimpleError -> HandshakeResult -> m () -> m ()
guardPeer _ peerId predicate err res action | predicate res peerId = do
  handshakeSend (Bai err :: HandshakeMessage c)
  throwError err
guardPeer _ _ _ _ action = action

peerIdMatch :: HandshakeResult -> NodeId -> Bool
peerIdMatch = (==) . hrPeerId
```

and later:

```
...
guardPeer (peerIdMatch)
guardPeer (not . peerIdMatch) 
```

Seems like a recurring pattern worth the refactoring? 🤔

Reply:
Yeah... but... it's different error constructors

Reply:
oh, I see, so you might need an injective type family to map one to the other to use it inside `guardPeer`? At which point it feels.. bulky I suppose?

Reply:
It's bulky, yeah, because that error type ought to not be thrown from another module, imho. But I'm a good 🏌️ , so here you go: ec2c276898ce0d261c8517187d01f08507be971b

Reply:
Good lad, thanks! 😁 ❤️